File: In/House/Q5/Solution.java
package In.House.Q5;

public class Solution {

    public String longestPalindrome(String s) {
        int len = s.length();
        if (len < 2) {
            return s;
        }
        int max = 1, idx = 0;
        boolean[][] dp = new boolean[len][len];
        for (int i = 0; i < len; ++i) {
            dp[i][i] = true;
        }
        char[] chars = s.toCharArray();
        for (int l = 2; l <= len; ++l) {
            for (int i = 0; i < len; ++i) {
                int j = i + l - 1;
                if (j >= len) {
                    break;
                }
                dp[i][j] = chars[i] != chars[j] ? false : (j - i < 3 ? true : dp[i + 1][j - 1]);
                if (!dp[i][j] || j - i + 1 <= max) {
                    continue;
                }
                max = j - i + 1;
                idx = i;
            }
        }
        return s.substring(idx, idx + max);
    }

    public static void main(String[] args) {
        Solution slt = new Solution();
        System.out.println(slt.longestPalindrome("babad")); // bab
        System.out.println(slt.longestPalindrome("cbbd")); // bb
        System.out.println(slt.longestPalindrome("a")); // a
        System.out.println(slt.longestPalindrome("ac")); // a
        System.out.println(slt.longestPalindrome("abb")); // bb
    }

}

File: In/House/Q5/solution.py

class Solution:
    def longestPalindrome(self, s: str) -> str:
        sLen = len(s)
        if sLen < 2:
            return s
        count = 1
        idx = 0
        dp = [[True for _ in range(sLen)] for _ in range(sLen)]
        chars = [c for c in s]
        for l in range(2, sLen + 1):
            for i in range(sLen):
                j = i + l - 1
                if j >= sLen:
                    break
                dp[i][j] = False if chars[i] != chars[j] else (True if j - i < 3 else dp[i + 1][j - 1])
                if dp[i][j] is False or j - i + 1 <= count:
                    continue
                count = j - i + 1
                idx = i
        return s[idx: idx + count]

slt = Solution()
print(slt.longestPalindrome('babad')) # bab
print(slt.longestPalindrome('cbbd')) # bb
print(slt.longestPalindrome('a')) # a
print(slt.longestPalindrome('ac')) # a
print(slt.longestPalindrome('abb')) # bb
print(slt.longestPalindrome('bb')) # bb




